exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,382 @@
+#ifdef _WIN32
+#define _CRT_SECURE_NO_WARNINGS
+#endif
+#include "sqlite.h"
+
+#include <inttypes.h>
+
+#include "assert.h"
+#include "cell.h"
+#include "exec.h"
+#include "nstring.h"
+#include "number.h"
+#include "object.h"
+#include "stack.h"
+
+#include "sqlite3.h"
+
+
+
+static Dictionary *Cursors;
+
+void sqlite_initModule(void)
+{
+    Cursors = dictionary_createDictionary();
+}
+
+void sqlite_shutdownModule(void)
+{
+    dictionary_freeDictionary(Cursors);
+}
+
+
+static Object *check_database(TExecutor *exec, Object *pdb)
+{
+    if (pdb == NULL || pdb->ptr == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        return NULL;
+    }
+    return pdb;
+}
+
+
+void object_releaseDatabaseObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                assert(sqlite3_close(o->ptr) == SQLITE_OK);
```

Comment:
Never ever put a statement with side effects like this inside `assert()`. What happens when you build without asserts enabled? The close function will never be called. Save the result in a variable and assert on the value of the result.

Reply:
Removed.  Also removed the assertion in general.  If the programmer fails to close the DB properly, then we cannot be responsible for trying to clean it up for them.